volumebinding: scheduler queueing hints - PersistentVolume

bells17 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

kube-scheduler implements scheduling hints for the VolumeBinding plugin.
The scheduling hints allow the scheduler to determine whether to retry or skip scheduling a Pod based on the changes made to the PersistentVolume resource referenced by the plugin.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/kubernetes/issues/118893
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/4247-queueinghint/README.md
Base PR: https://github.com/kubernetes/kubernetes/pull/124939

#### Special notes for your reviewer:

<details><summary>Fields Impacting QueueingHintFn</summary>
<p>
<strong>PersistentVolume (PV)</strong> is not included in this table because it can undergo extensive changes when a conversion is performed by csi-translation-lib.
</p>

| resource | field | Referenced in PreFilter+Filter? | Admission Overwrite Prevention Config | Need to Check Changes in QHint? |
| ---- | ---- | ---- | --- | --- |
| PersistentVolume(PV) | .metadata.labels  | o | x | o |
| PersistentVolume(PV) | .metadata.annotations  | o | x | o |
| PersistentVolume(PV) | .spec.capacity | o | x | o |
| PersistentVolume(PV) | .spec.(PersistentVolumeSource) | o | o | x |
| PersistentVolume(PV) | .spec.accessModes | o | x | o |
| PersistentVolume(PV) | .spec.claimRef | o | x | o |
| PersistentVolume(PV) | .spec.persistentVolumeReclaimPolicy | o | x | o |
| PersistentVolume(PV) | .spec.storageClassName | o | x | o |
| PersistentVolume(PV) | .spec.mountOptions | o | x | o |
| PersistentVolume(PV) | .spec.volumeMode | o | o | x |
| PersistentVolume(PV) | .spec.nodeAffinity | o | o | o |
| PersistentVolume(PV) | .spec.volumeAttributesClassName | o | o | x |
| PersistentVolume(PV) | .status.phase | o | x | o |
| PersistentVolume(PV) | .status.message | o | x | o |
| PersistentVolume(PV) | .status.reason | o | x | o |
| PersistentVolume(PV) | .status.lastPhaseTransitionTime | o | x | o |

ref(ja): https://zenn.dev/bells17/scraps/65bd6891012bdc

</details> 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-scheduler implements scheduling hints for the VolumeBinding plugin.
The scheduling hints allow the scheduler to retry scheduling a Pod that was previously rejected by the VolumeBinding plugin only if a new resource referenced by the plugin was created or an existing resource referenced by the plugin was updated.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bells17 committed:
volumebinding: scheduler queueing hints - PersistentVolume

bells17 commented:
/cc @sanposhiho @utam0k

sanposhiho commented:
/kind feature

AxeZhan commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -99,7 +100,7 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
 		// We bind PVCs with PVs, so any changes may make the pods schedulable.
 		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeAddOrChange},
```
nit
```suggestion
		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeChange},
```
I think this is enough, let's keep the same behaviour with other plugins.

bells17 replied:
I fixed it.

AxeZhan commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -170,6 +171,29 @@ func (pl *VolumeBinding) podHasPVCs(pod *v1.Pod) (bool, error) {
+		}
+	}
+
+	logger.V(4).Info("PersistentVolumeClaim was created or updated, but it doesn't make this pod schedulable")
```
```suggestion
	logger.V(4).Info("PersistentVolume was created or updated, but it doesn't make this pod schedulable")
```

bells17 replied:
I fixed it.

bells17 committed:
Fixed points mentioned in the review

bells17 committed:
Fixed points mentioned in the review

bells17 commented:
/retest

bells17 commented:
@AxeZhan I've addressed the points you mentioned in your feedback. I would appreciate it if you could take another look at the changes.
https://github.com/kubernetes/kubernetes/pull/124960/commits/9d537e43e0b24ff131e4d11e6560d271adf2279a

AxeZhan commented:
/cc @carlory 
I'm not familiar with storage😅. I'll let folks from sig-storage take a look first.

carlory commented:
I'm quite busy today, I'll find time to take a look this week

bells17 commented:
/cc @jsafrane @xing-yang

sanposhiho commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -170,6 +171,29 @@ func (pl *VolumeBinding) podHasPVCs(pod *v1.Pod) (bool, error) {
+	)
+
+	for _, vol := range pod.Spec.Volumes {
+		if vol.Name == newPV.Name {
```
How is `pod.Spec.Volumes[*].Name` related to PV's name?

bells17 replied:
Sorry, that was my mistake, so I've fixed it.

bells17 committed:
Fixed points mentioned in the review

bells17 commented:
@sanposhiho  Thank you for your review. I have made the necessary changes, so please take another look.

carlory commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -99,7 +100,7 @@ func (pl *VolumeBinding) EventsToRegister() []framework.ClusterEventWithHint {
 		// We bind PVCs with PVs, so any changes may make the pods schedulable.
 		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeChange},
```
I don't know why this plugin registers PV events. IMO, The pv events can not directly influence a pod's schedulability. During the process of binding a PV to a PVC, the PV are always updated first, then the PVC is updated. So the later PVC event may make the Pod become schedulable. so registering PV events maybe not necessary but it may accelerate the pod scheduling process (it doesn't wait for the pvc to be bound and put the pod into backoffQ).

It was added by this PR: https://github.com/kubernetes/kubernetes/pull/100003

cc @jsafrane @sanposhiho @cofyc @Huang-Wei @saad-ali

bells17 commented:
/close
https://github.com/kubernetes/kubernetes/pull/124961#discussion_r1620473466

bells17 commented:
/reopen

bells17 commented:
/close
